<launch>

    <arg name="enable_debug"                default="false"/>
    <arg name="is_ground_vehicle"           default="true"/>
    <arg name="enable_wf_control"           default="true"/>
    <arg name="enable_sf_control"           default="false"/>
    <arg name="enable_attractor_control"    default="false"/>
    <arg name="enable_command_weighting"    default="false"/>

    <arg name="total_horiz_scan_points"     default="720"/>
    <arg name="num_horiz_scan_points"       default="360"/>
    <arg name="horiz_scan_start_index"      default="180"/>
    <arg name="horiz_scan_limit"            default="1.5707"/>
    <arg name="horiz_sensor_max_distance"   default="0.1"/>
    <arg name="horiz_sensor_max_distance"   default="40.0"/>
    <arg name="horiz_sensor_min_noise"      default="0.1"/>
    <arg name="reverse_horiz_scan"          default="true"/>

    <arg name="enable_safety_boundary"      default="true"/>
    <arg name="enable_safety_box"           default="true"/>
    <arg name="front_safety_distance"       default="0.25"/>
    <arg name="side_safety_distance"        default="0.35"/>
    <arg name="safety_radius"               default=".5"/>
    <arg name="close_side_speed"            default="0.05"/>

    <node pkg="nearness_control" type="nearness_controller_node" name="nearness_control_node" output="screen">

        <!-- SUBSCRIBED TOPICS -->
        <remap from="nearness_controller/horiz_scan" to="scan"/>
        <remap from="nearness_controller/vert_scan" to="scan1"/>
        <remap from="nearness_controller/next_waypoint" to="lookahead_point"/>
        <remap from="nearness_controller/odometry" to="odometry_map"/>
        <remap from="nearness_controller/joy" to="joy"/>
        <remap from="nearness_controller/estop_cmd" to="estop_cmd"/>
        <remap from="nearness_controller/sim_control_commands" to="nearness_cmd_vel"/>

        <!-- Enable switches -->
        <param name="enable_debug" type="bool" value="$(arg enable_debug)"/>
        <param name="is_ground_vehicle" type="bool" value="$(arg is_ground_vehicle)"/>
        <param name="enable_wf_control" type="bool" value="$(arg enable_wf_control)"/>
        <param name="enable_sf_control" type="bool" value="$(arg enable_sf_control)"/>
        <param name="enable_attractor_control" type="bool" value="$(arg enable_attractor_control)"/>
        <param name="enable_command_weighting" type="bool" value="$(arg enable_command_weighting)"/>

        <!-- Sensor Parameters -->
        <param name="total_horiz_scan_points" type="int" value="$(arg total_horiz_scan_points)"/>
        <param name="num_horiz_scan_points" type="int" value="$(arg num_horiz_scan_points)"/>
        <param name="horiz_scan_start_index" type="int" value="$(arg horiz_scan_start_index)"/>
        <param name="horiz_scan_limit" type="double" value="$(arg horiz_scan_limit)"/>
        <param name="scan_start_location" type="string" value="back" /> <!-- Valid args: 'forward', 'back', 'left', 'right' -->
        <param name="horiz_sensor_min_distance" type="double" value="$(arg horiz_sensor_min_distance)"/>
        <param name="horiz_sensor_max_distance" type="double" value="$(arg horiz_sensor_min_distance)"/>
        <param name="horiz_sensor_min_noise" type="double" value="$(arg horiz_sensor_min_noise)"/>
        <param name="reverse_horiz_scan" type="bool" value="$(arg reverse_horiz_scan)"/>
        <param name="num_horiz_fourier_terms" type="int" value="5"/>

        <param name="total_vert_scan_points" type="int" value="1081"/>
        <param name="num_vert_scan_points" type="int" value="720"/>
        <param name="vert_scan_limit" type="double" value="1.5707"/>
        <param name="scan_start_location" type="string" value="back" /> <!-- Valid args: 'forward', 'back', 'left', 'right' -->
        <param name="vert_sensor_min_distance" type="double" value="0.1"/>
        <param name="vert_sensor_max_distance" type="double" value="25.0"/>
        <param name="vert_scan_start_index" type="int" value="180"/>
        <param name="vert_sensor_min_noise" type="double" value="0.1"/>
        <param name="reverse_vert_scan" type="bool" value="false"/>

        <!-- Safety Parameters -->
        <param name="enable_safety_boundary" type="bool" value="$(arg enable_safety_boundary)"/>
        <param name="enable_safety_box" type="bool" value="$(arg enable_safety_box)"/>
        <param name="front_safety_distance" type="double" value="$(arg front_safety_distance)"/>
        <param name="side_safety_distance" type="double" value="$(arg side_safety_distance)"/>
        <param name="safety_radius" type="double" value="$(arg safety_radius)"/>
        <param name="close_side_speed" type="double" value="$(arg close_side_speed)"/>

        <!-- Parameters for Forward Speed Controller -->
        <param name="forward_speed_k_hb_1" type="double" value=".5"/>
        <param name="forward_speed_k_hb_2" type="double" value=".5"/>
        <param name="forward_speed_k_vb_1" type="double" value="0.25"/>
        <param name="forward_speed_k_vb_2" type="double" value="0.5"/>
        <param name="forward_speed_min" type="double" value="0.1"/>
        <param name="forward_speed_max" type="double" value=".25"/>

        <!-- Parameters for WF Yaw Rate Control -->
        <param name="yaw_rate_k_hb_1" type="double" value="1.0"/>
        <param name="yaw_rate_k_hb_2" type="double" value="2.0"/>  <!-- -2 -->
        <param name="yaw_rate_max" type="double" value=".1"/>

        <!-- Parameters for SF Yaw Rate Control -->
        <param name="h_sf_k_thresh" type="double" value="3.0"/>
        <param name="h_sf_k_0" type="double" value="0.4"/>
        <param name="h_sf_k_psi" type="double" value="0.5"/>
        <param name="h_sf_k_d" type="double" value="0.5"/>

        <!-- Parameters for Attractor Yaw Rate Control -->
        <param name="yaw_rate_k_att_0" type="double" value="-.5"/>
        <param name="yaw_rate_k_att_d" type="double" value="1.0"/>

        <!-- Additional parameters for aerial vehicles -->
        <param name="lateral_speed_k_hb_1" type="double" value="4.0"/>
        <param name="lateral_speed_max" type="double" value="2.0"/>

        <param name="vert_speed_k_vb_1" type="double" value="-3.0"/>
        <param name="vert_speed_k_vb_2" type="double" value="-1.0"/>  <!-- -2 -->
        <param name="vert_speed_max" type="double" value="2.0"/>

    </node>
  <!--  <node pkg="topic_tools" name="control_relay" type="relay" args="/nearness_controller/sim_control_commands /cmd_vel" /> -->

</launch>
